Virtual Routes Support

illBeRoy opened the pull request:
<!--
  PULL REQUEST TEMPLATE
  ---
  Please use English language
  Please don't delete this template
  ---
  Update "[ ]" to "[x]" to check a box in any list below.
  ---
  To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.
-->

## **Summary**
This PR introduces the _**Virtual Routes**_ feature, which allows users to introduce dynamically created content into their docsify websites. For instance:

```js
window.$docsify = {
  routes: {
    '/items/(.+)': function (route, matched) {
      return `
        # Item Page: ${matched[1]}
        This is an item
      `;
    }
  }
}
```

Will make docsify handle every route under `/items/*` and generate a dynamic markdown using a predefined string, a function or an async function.

For a full specification of the feature, please see the relevant issue: [docsifyjs/docsify#1737](https://github.com/docsifyjs/docsify/issues/1737#issuecomment-1058633104)

<!--
 THIS IS REQUIRED! Please describe what the change does and why it should be merged.
-->

<!--
  If changing the UI in any way, please provide the a **before/after** screenshot:
-->

## **What kind of change does this PR introduce?**

###  Feature: "routes" configuration option
In order to implement the feature, the following changes were made:
1. Created a new mixin called `VirtualRoutes` (under `src/core/virtual-routes`)
2. Updated the `Fetch` mixin: the method `_fetch` now attempts to match virtual routes before making a request
3. Created a new E2E test suite under `test/e2e/virtual-routes.test.js`, which covers: route types, regex matching and order of handling
4. Had to updated a test in `test/e2e/configuration.test.js` due to a playwright bug (more about it in the next section)

<!--
  Copy/paste one of the following options:
-->

<!--
  Bugfix
  Feature
  Code style update
  Refactor
  Docs
  Build-related changes
  Repo settings
  Other
-->

<!--
  If you chose Other, please describe.
-->

## **For any code change,**

- [x] Related documentation has been updated if needed
- [x] Related tests have been updated or tests have been added

## **Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

## **Related issue, if any:**

https://github.com/docsifyjs/docsify/issues/1737

<!-- Paste issue's link or number hashtag here. -->

## **Tested in the following browsers:**

- [x] Chrome
- [x] Firefox
- [x] Safari
- [ ] Edge
- [ ] IE

jhildenbiddle commented:
This looks really great, @illBeRoy. I'll take a closer look and provide a review as soon as possible.

Looking forward to getting this merged! Thanks again!

jhildenbiddle reviewed (changes requested):
This is very well done PR. A few minor doc tweaks and two code-related issues, but overall really great and nearly merge ready. Thanks again!

illBeRoy commented on src/core/fetch/index.js:
```diff
@@ -96,13 +96,32 @@ export function Fetch(Base) {
         this.isHTML = /\.html$/g.test(file);
 
         // Load main content
+        const req = Promise.resolve()
```
This is perfectly doable. There are two options:
1. Rewrite without promises - might make the code harder to read for modern JS contributors
2. Use a stand-in polyfill, like [promise-polyfill](https://www.npmjs.com/package/promise-polyfill) - in which case the refactor is going to be 0 work, but it will add up to 1kb to the bundle.

WDYT?

jhildenbiddle replied:
I'm not a big fan of either option, to be honest. 😉 

If I had to choose one, I'd probably opt to rewrite the code without promises since Docsify includes no other polyfills and we're talking about a total of ~30 lines here. We can worry about modernizing this block and other areas of our codebase if/when we bump to v5 and drop legacy browser support (🤞).

trusktr replied:
Can we just drop legacy support (browsers without Promise are _really really_ old) and bump to V5, without requiring that V5 also had a bunch of feature changes too? We could always bump to V6 for that, but dropping legacy browsers will really help simplify code.

illBeRoy commented on test/helpers/navigate.js:
```diff
@@ -0,0 +1,12 @@
+  await mainElm.waitForElementState('stable');
+}
+
+module.exports.navigateToRoute = navigateToRoute;
```
First of all, you are correct, there's no need to await `#main` to be stable. I'll remove that.

The reason I did not use `page.goto` is that our test server does not support appending hashes to it, and returns 404. Instead, I opted to use `page.evaluate(() => (location.hash = ...))`. `navigateToRoute` is just a simple wrapper around that, meant to serve as a meaningful alias conveying the action's intent.

Should I update it with an optional third parameter that tells which type of routing we use? e.g.:
```js
async function navigateToRoute(page, route, { routerMode = 'hash' } = {}) {
  if (routerMode === 'hash') {
    await page.evalulate(() => location.hash = ... );
  } else {
    await page.evalulate(() => history.pushState(...));
  }
}
```

wdyt?

jhildenbiddle replied:
We can get the `routerMode` via the DOM instead of requiring devs to specify it as an argument.

Also, after looking at the `page.goto()` [options](https://playwright.dev/docs/api/class-page#page-goto) more carefully, I realized most of them wouldn't apply because we aren't actually navigating to a new page. The most interesting option allows us to wait for `networkidle` which we can do via `page.waitForLoadState()`:

```js
async function navigateToRoute(page, route) {
  const routerMode = await page.evaluate(() => window.$docsify.routerMode);

  if (routerMode === 'history') {
    await page.evaluate(() => history.pushState(...));
  } else {
    await page.evaluate(r => (window.location.hash = r), route);
  }

  await page.waitForLoadState('networkidle');
}
```

How does that look?

illBeRoy replied:
Looks good 👍

jhildenbiddle replied:
Any reason why we can't get the `routerMode` via `page.evaluate()` instead of requiring the value to be passed as an argument to `navigateToRoute()`?

From example above:

```js
async function navigateToRoute(page, route) {
  const routerMode = await page.evaluate(() => window.$docsify.routerMode);
```

illBeRoy replied:
haven't noticed that in your example :) changed

illBeRoy commented:
Hey! I updated the PR according to your notes.

I refactored the code and removed all usage of promises. I hope that the code is still clear enough, and would appreciate your feedback

jhildenbiddle commented on docs/configuration.md:
````diff
@@ -48,6 +48,8 @@ window.$docsify = {
 };
 ```
 
+Do note that order matters! If a route can be matched by multiple aliases, the one you declared first takes precedence.
````
Can we move this above the code example?

jhildenbiddle commented on docs/configuration.md:
````diff
@@ -680,6 +682,113 @@ window.$docsify = {
+}
+```
+
+Do note that order matters! Routes are matched the same order you declare them in, which means that in cases where you have overlapping routes, you might want to list the more specific ones first:
````
As above, let's move this above the first code example.

jhildenbiddle commented on src/core/virtual-routes/exact-match.js:
```diff
@@ -0,0 +1,17 @@
+ * @returns {string}
+ */
+export function makeExactMatcher(matcher) {
+  const matcherWithBeginningOfInput = matcher.startsWith('^')
```
Method `startsWith()` not supported by legacy browsers.

jhildenbiddle replied:
👍 Working in IE11 after fix.

jhildenbiddle commented on src/core/virtual-routes/exact-match.js:
```diff
@@ -0,0 +1,17 @@
+    : `^${matcher}`;
+
+  const matcherWithBeginningAndEndOfInput =
+    matcherWithBeginningOfInput.endsWith('$')
```
Method `endsWith()` not supported by legacy browsers.

illBeRoy replied:
took care of this. btw, how do you check for this? do you have a tool \ lint rule for that?

jhildenbiddle replied:
Nope. Wish I did. 

I use a free [BrowserStack](https://www.browserstack.com) account to manually test in IE11. Browserstack has a nice [open source program](https://www.browserstack.com/open-source) that allows open source authors to use a limited set of their tools for free. Super handy.

jhildenbiddle replied:
FYI, I noticed we have an `endsWith()` function here:

https://github.com/docsifyjs/docsify/blob/b8b221f3a87652acf6d57286004d711a3671537f/src/core/router/util.js#L117-L119

Not saying you have to use that, but you could add a `startsWith()` function in this same file and leverage those if you preferred.

jhildenbiddle commented on test/e2e/virtual-routes.test.js:
```diff
@@ -0,0 +1,284 @@
+const docsifyInit = require('../helpers/docsify-init');
+const { navigateToRoute } = require('../helpers/navigate');
+const { test, expect } = require('./fixtures/docsify-init-fixture');
+
```
Can we add a basic test using `routerMode: 'history'` to verify that routes and the `navigateToRoute()` helper works as expected?

illBeRoy replied:
ok so seems like `docsifyInit` cannot create test pages that use `routerMode: history`:
![image](https://user-images.githubusercontent.com/6681893/169371509-a7ae3243-4607-45c3-b433-16c07d7e902f.png)
![image](https://user-images.githubusercontent.com/6681893/169371604-27434437-9edf-41a0-81de-76ea9895033e.png)

I can fix it, but IMO it seems a little out of scope for this PR. How do we feel about moving the `navigateToPage` helper into the `virtual-routes.test.js` file and explicitly supporting only `hash` mode?

jhildenbiddle replied:
Agreed on fixing `docsifyInit()` being out of scope.

Also agree on the simple solution: move the `navigateToRoute()` helper into the `virtual-routes.test.js`.

However, if you are motivated to do so you could verify that `navigateToRoute()` works as expected on docsify sites using `routeMode: 'history'` when `docsifyInit()` isn't used. Assuming this work, this would allow the helper to remain in the `/test/helpers` directory so it can be used today when testing against the `/docs/` site or in the future if/when we fix `docsifyInit()`.

I gave this a shot myself by setting `routerMode: 'history'` in `/docs/index.html` and using the following test (WIP):

```js
test('routerMode:history', async ({ page }) => {
  // Docs site
  page.goto('http://localhost:3001/docs/');

  // Wait for docsify to initialize
  await page.waitForSelector('#main > *');
  await page.waitForLoadState('networkidle');

  // Verify routerMode
  const $docsify = await page.evaluate(() => window.$docsify);
  expect($docsify).toHaveProperty('routerMode', 'history');

  // URL Before
  console.log(page.url());

  // Navigate
  navigateToRoute(page, '/quickstart');

  // URL Before
  console.log(page.url());
});
```

The above test did not work for me for two reasons:

1. First, the `route` wasn't being passed to `page.evaluate` when the `routeMode` was set to `history`. Fixed that (below).
   ```js
   // Before
   if (routerMode === 'history') {
     await page.evaluate(r => window.history.pushState({ key: r }, '', r));
   }

   // After
   if (routerMode === 'history') {
     await page.evaluate(
       r => window.history.pushState({ key: r }, '', r),
       route
     );
   }
   ```
1. After calling `history.pushState`, the I received the following Playwright error:
   ```
   page.evaluate: Target closed
   ```

I started down the rabbit hole and began playing `page.waitForNavigation()` and researching how to use `history.pushState` without closing the Playwright `page` object, then I realized I was going down the rabbit hole and stopped myself. :smile:

Anyway, no need to go down this rabbit hole yourself if you aren't interested. Just sharing my results after tinkering for a bit in case you wanted to continue the effort.

illBeRoy replied:
Interesting. Perhaps pushing history state is considered by the driver as navigation, therefore destroying the context in which the `evaluate` method was called. I'll try to take a look there; in the meantime, I moved the function into the test file :)

illBeRoy commented:
Attended all comments :) LMK what you think

jhildenbiddle commented on test/e2e/virtual-routes.test.js:
```diff
@@ -0,0 +1,293 @@
+  });
+});
+
+/**
```
Last nitpick (promise)...

Can we move this after the imports and before the first `test()` block? This has been the general pattern for local helpers in other test files.

illBeRoy replied:
np, done & done

jhildenbiddle reviewed (dismissed):
Looks great. Excited for this. Thanks, @illBeRoy!

trusktr commented on docs/configuration.md:
```diff
@@ -680,6 +681,95 @@ window.$docsify = {
+      if (matched[0] === 'cats') {
+        next();
+      } else {
+        try {
```
I personally would rather show new syntax that works reasonably in all current browsers, and let people who _need_ to support older browsers figure what old syntax to use. That's just my preference.

We're just making our (limited time maintainer) lives more difficult for a very small minority of users.

trusktr replied:
I personally believe there's not much sense in not teaching people async/await these days. It has been _7 years_ since ES2015.

illBeRoy replied:
@jhildenbiddle wdyt?

jhildenbiddle replied:
@trusktr makes some valid points. Here's my $0.02...

> Note that async tasks that throw errors without using async/await syntax, will not be caught by the try-catch block.

Yep. Good catch (pun intended). I think I was a little hyper-focused on showing a simple example with try/catch block for async tasks to prevent uncaught errors from breaking sites and missed the fact that my simple example doesn't actually work. 🤦‍♂️ 

> I personally would rather show new syntax that works reasonably in all current browsers, and let people who need to support older browsers figure what old syntax to use. That's just my preference.

Let's accommodate both scenarios by showing an example that uses an ES6+ feature that easily be polyfilled by those who care about legacy browser support:

```js
// RegEx match w/ asynchronous function
'/baz/(.*)': function(route, matched, next) {
  // Requires `fetch` polyfill for legacy browsers (https://github.github.io/fetch/)
  fetch('/api/users?id=12345')
    .then(function(response) {
      return response.text()
    })
    .then(function(body) {
      next('# Custom Markdown');
    })
    .catch(function(ex) {
      console.log('parsing failed', ex)
    });
}
 ```

The alternative is to simplify the example as much as possible and just focus on the addition of the `next` function for async tasks:

```js
// RegEx match w/ asynchronous function
'/baz/(.*)': function(route, matched, next) {
  // Async task(s)...
  next('# Custom Markdown');
}
 ```

I'm fine either way. I think I prefer the first option for this scenario because fetching remote data is a real-world use case for virtual routes.

illBeRoy replied:
NP, updated the docs :)

jhildenbiddle replied:
@trusktr Sound good to you?

trusktr reviewed (changes requested):
Looks very good, but just one nit pick on example code for legacy browsers.

trusktr commented:
I approved, I like the example with link to polyfill if needed. An more modern user will pick up on the opportunity for async/await.
